lib/posix/pthread.c

```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
  *
  * See IEEE 1003.1
  */
-int pthread_create(pthread_t *newthread, const pthread_attr_t *_attr,
-		   void *(*threadroutine)(void *), void *arg)
+int pthread_create(pthread_t *th, const pthread_attr_t *_attr, void *(*threadroutine)(void *),
+		   void *arg)
 {
-	int rv;
-	int32_t prio;
+	int err;
 	k_spinlock_key_t key;
-	uint32_t pthread_num;
-	k_spinlock_key_t cancel_key;
-	pthread_condattr_t cond_attr;
-	struct posix_thread *thread;
-	const struct pthread_attr *attr = (const struct pthread_attr *)_attr;
+	struct posix_thread *t;
+	pthread_barrier_t barrier;
+	struct pthread_attr attr;
 
-	/*
-	 * FIXME: Pthread attribute must be non-null and it provides stack
-	 * pointer and stack size. So even though POSIX 1003.1 spec accepts
-	 * attrib as NULL but zephyr needs it initialized with valid stack.
-	 */
-	if ((attr == NULL) || (attr->initialized == 0U)
-	    || (attr->stack == NULL) || (attr->stacksize == 0)) {
+	if (!pthread_attr_is_valid((const struct pthread_attr *)_attr)) {
 		return EINVAL;
 	}
 
-	key = k_spin_lock(&pthread_pool_lock);
-	for (pthread_num = 0;
-	    pthread_num < CONFIG_MAX_PTHREAD_COUNT; pthread_num++) {
-		thread = &posix_thread_pool[pthread_num];
-		if (thread->state == PTHREAD_EXITED || thread->state == PTHREAD_TERMINATED) {
-			thread->state = PTHREAD_JOINABLE;
-			break;
+	/* copy the attribute in to the */
+	(void)memcpy(&attr, _attr, sizeof(attr));
+
+	key = k_spin_lock(&q_lock);
+	/* recycle old nodes if they are available */
+	do {
+		struct posix_thread *safe_t;
+
+		SYS_DLIST_FOR_EACH_CONTAINER_SAFE(&done_q, t, safe_t, q_node) {
+			if (t->detachstate == PTHREAD_CREATE_JOINABLE) {
+				/* thread has not been joined yet */
+				continue;
+			}
+			sys_dlist_remove(&t->q_node);
+			sys_dlist_append(&ready_q, &t->q_node);
+			t->qid = POSIX_THREAD_READY_Q;
 		}
+	} while (false);
+
+	/* check if a thread is available to spawn */
+	t = NULL;
+	if (!sys_dlist_is_empty(&ready_q)) {
+		t = CONTAINER_OF(sys_dlist_get(&ready_q), struct posix_thread, q_node);
```

Comment:
Should the thread be pulled out of the ready_q right here? Seems like there's a possible race with two pthread_create calls both ending up with the same thread?

Reply:
I don't think that's possible - in order to manipulate the which queue a thread is in (i.e. `q_node` within each `struct posix_thread`) `q_lock` needs to be held.

It's clearly held here for the entirety of the queue manipulation.

Reply:
q_lock is held here, I'm wondering why you're setting qid to `POSIX_THREAD_NONE_Q` without also removing `t` from `ready_q`. That seems like a bug.

Reply:
Especially when you've got a failure case down below that re-inserts it into `ready_q` and resets qid to `POSIX_THREAD_READY_Q`.

Reply:
Thanks, that answers my concern.